source/common/upstream/ring_hash_lb.cc

```diff
@@ -60,34 +62,108 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
+      config && config->has_replication_factor()
+          ? absl::optional<uint64_t>(config->replication_factor().value())
+          : absl::nullopt;
+
+  // Sanity-check ring size bounds.
+  if (min_ring_size > max_ring_size) {
+    throw EnvoyException(fmt::format("ring hash: minimum_ring_size ({}) > maximum_ring_size ({})",
+                                     min_ring_size, max_ring_size));
+  }
+
+  // Compute the "effective" weight of each host as the product of its own weight and the locality
+  // weight, if given. Sum these effective weights.
+  uint64_t weighted_sum = 0;
+  std::unordered_map<HostConstSharedPtr, uint32_t> effective_weights;
+  for (uint64_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      auto host_weight = host->weight();
+      ASSERT(host_weight != 0);
+      // NOTE: Locality weight may be zero. In this case, all hosts in the locality are assigned
+      //       no load. When we build the ring below, such hosts will have no entries in the ring.
+      // TODO: When we move to C++17, change this to `locality_weights[i]` (i.e. use
+      //       std::shared_ptr::operator[]) rather than dereferencing locality_weights explicitly.
+      auto locality_weight = locality_weights == nullptr ? 1 : (*locality_weights)[i];
+      auto effective_weight = host_weight * locality_weight;
+      weighted_sum += effective_weight;
+      effective_weights[host] = effective_weight;
+    }
+  }
+
+  // We can't do anything sensible with no hosts.
+  if (weighted_sum == 0) {
+    return;
+  }
 
-  uint64_t hashes_per_host = 1;
-  if (hosts.size() < min_ring_size) {
-    hashes_per_host = min_ring_size / hosts.size();
-    if ((min_ring_size % hosts.size()) != 0) {
-      hashes_per_host++;
+  // Determine the valid range of replication factors given the weighted sum and ring size bounds,
+  // and the actual replication factor within that range. This may differ from the configured
+  // replication factor if provided. Note that we use floating point math here and round up when
+  // adding hosts to the ring, rather than trying to preserve exact integer ratios. This is good
+  // enough, since the distribution of hosts around the ring wouldn't exactly match the given
+  // weights anyway, even with a good hash function.
+  const double min_replication_factor =
+      static_cast<double>(min_ring_size) / static_cast<double>(weighted_sum);
+  const double max_replication_factor =
+      static_cast<double>(max_ring_size) / static_cast<double>(weighted_sum);
+  double replication_factor;
+  if (configured_replication_factor) {
+    // If the configuration includes a replication factor, try to use it and warn if not possible.
+    replication_factor = static_cast<double>(configured_replication_factor.value());
+    if (replication_factor < min_replication_factor) {
+      ENVOY_LOG(warn, "ring hash: replication_factor ({}) too small for minimum_ring_size ({})",
```

Comment:
I think we could just simplify this and just write `replication_factor = std::min(max_replication_factor, std::max(min_replication_factor, configured_replication_factor))`. The warnings are maybe too strong because it's normal operational behavior to hit the max factor and then tradeoff some granularity.

Reply:
Not sure... I think it's good to warn the user if the `replication_factor` they specified isn't being used, because they have no other way of knowing they specified something Envoy can't respect. Maybe they don't have decent ring size bounds, maybe they don't have a reasonable replication factor, maybe their weights are weird or the number of healthy hosts isn't what they'd expect...

I don't feel extremely strongly about it, though. I mean, you're right that we are going to some effort to make sure that things behave reasonably when you have fractional replication factors. I'm mostly just worried about the case where replication factors start to get very low and the request distribution gets messed up as a result.

Reply:
Yeah, my main concern is that warnings are generally not super helpful in production, since they aren't actionable and require human interpretation. If you think this is serious enough of a condition to warrant monitoring, then we probably should have some stats monitoring. 

How about just exposing the effective replication factor as a gauge?  Then you can have a monitoring alert if this becomes too low.

Reply:
I like the gauge idea a lot. I'll poke around the codebase and see if I can figure out how to do that.

Reply:
@htuch, latest push introduces a new stats scope, `ring_hash_lb` as a child of the cluster stats scope passed in, with `size` and `replication_factor` gauges. Thanks for the suggestion!